src/hmem_synapseai.c

```diff
@@ -37,6 +37,155 @@
 #include "ofi_hmem.h"
 #include "ofi.h"
 
+#if HAVE_SYNAPSEAI
+
+#include <dlfcn.h>
+#include "habanalabs/synapse_api.h"
+
+struct synapseai_ops {
+	int (*hcclLookupDMABuff)(uint64_t addr, uint64_t size, int* fd);
+};
+
+static void *synapseai_handle;
+static struct synapseai_ops synapseai_ops;
+
+/**
+ * @brief Initialize synapseai dl handle
+ * 
+ * @return On success, return 0. On failure, return
+ * a negative error code.
+ */
+static int synapseai_dl_init(void)
+{
+	synapseai_handle = dlopen("libSynapse.so", RTLD_NOW);
+	if (!synapseai_handle) {
+		FI_INFO(&core_prov, FI_LOG_CORE,
+			"Failed to dlopen libSynapse.so\n");
+		return -FI_ENOSYS;
+	}
+
+	synapseai_ops.hcclLookupDMABuff = dlsym(synapseai_handle, "hcclLookupDMABuff");
+	if (!synapseai_ops.hcclLookupDMABuff) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"Failed to find hcclLookupDMABuff\n");
+		goto err;
+	}
+
+	return FI_SUCCESS;
+err:
+	dlclose(synapseai_handle);
+	return -FI_ENODATA;
+}
+
+/**
+ * @brief dlclose the synapseAI handle.
+ * 
+ */
+static void synapseai_dl_cleanup(void)
+{
+	dlclose(synapseai_handle);
+}
+
+/**
+ * @brief Initialize SynapseAI hmem interface by dlopen the SynapseAI library
+ * 
+ * @return On success, return 0. On failure, return negative error code.
+ */
+int synapseai_init(void)
+{
+	int ret;
+
+	ret = synapseai_dl_init();
+	if (ret != FI_SUCCESS)
+		return ret;
+
+	return FI_SUCCESS;
+}
+
+/**
+ * @brief Clean up SynapseAI intereface (dlclose)
+ * @return return 0. 
+ */
+int synapseai_cleanup(void)
```

Comment:
Similarly, can we combine `synapseai_cleanup` and `synapseai_dl_cleanup`

Reply:
Done